URL matcher

ivanprado opened the pull request:
Read the documentation before reviewing the code to have a better context.

BurnzZ commented on url_matcher/matcher.py:
```diff
@@ -0,0 +1,174 @@
+
+    def remove(self, id: Any):
+        if id in self.patterns:
+            patterns = self.patterns[id]
```
_**optional**_: We could refactor this into the following to remove code indentation:

```python
patterns = self.patterns.get(id)
if not patterns:
    return
```

ivanprado replied:
Good idea. Done

BurnzZ commented on url_matcher/matcher.py:
```diff
@@ -0,0 +1,174 @@
+    priority: int = 500
+
+    def get_domains(self) -> List[str]:
+        return [domain for pattern in self.include if (domain := get_pattern_domain(pattern))]
```
We're using the `:=` walrus operator here but support a Python version as low as **3.6**. This would fail tests. 

~Also a bit curious if the coverage reports this one being touched by tests?~ nvm, tox was failing in on **3.6**: https://github.com/zytedata/url-matcher/runs/4289469410?check_suite_focus=true#step:5:239

ivanprado replied:
Yeah, I'll fix it. Thank you!

BurnzZ commented on url_matcher/patterns.py:
```diff
@@ -0,0 +1,265 @@
+        return True
+
+
+def _path_or_fragment_re(path_or_fragment: str) -> Pattern:
```
_**optional:**_ Might as well be a static method of the `PatternMatcher` class.

ivanprado replied:
Addressed

BurnzZ reviewed (commented):
Performed an initial review here @ivanprado and I must say, well done on creating such a clean and well-structured package! 🙌

ivanprado commented on .flake8:
```diff
@@ -0,0 +1,20 @@
+    # Similar to above, let black take care of the formatting.
+    W503,
+
+    # Refers to "necessary dict call - rewrite as a literal".
```
Addressed!

ivanprado commented on README.rst:
```diff
@@ -0,0 +1,87 @@
+Using sphinx-autobuild
+~~~~~~~~~~~~~~~~~~~~~~
+
+When working on documentation,it is convenient to use sphinx-autobuild.
```
Addressed!

ivanprado commented on docs/intro.rst:
```diff
@@ -0,0 +1,185 @@
+pattern ``example.com`` describes any URL whose domain is ``example.com``
+or any of its subdomains.
+
+A single patter is sometimes not enough to describe which URLs to match.
```
Addressed

ivanprado commented on requirement-dev.txt:
```diff
@@ -0,0 +1,2 @@
+pre-commit
```
Addressed!

ivanprado commented on setup.py:
```diff
@@ -0,0 +1,45 @@
+        "Programming Language :: Python :: 3.7",
+        "Programming Language :: Python :: 3.8",
+        "Programming Language :: Python :: 3.9",
+    ],
```
Addressed!

ivanprado commented on url_matcher/matcher.py:
```diff
@@ -0,0 +1,174 @@
+    include: List[str]
+    exclude: List[str] = field(default_factory=list)
+    priority: int = 500
+
```
This method is only invoked only once when a new rule is added. So the overhead is very small. Let's leave this optimisation for the future then.

ivanprado commented:
Thank you @BurnzZ for your review, I've addressed your comments. I'm going to merge this PR and create a new one that can be reviewed properly (sorry for the mess with the files)
